frame/airdrop/src/lib.rs

```diff
@@ -0,0 +1,977 @@
+		}
+
+		/// Stop an Airdrop.
+		///
+		/// Returns the amount of unclaimed funds from the airdrop upon success.
+		///
+		/// # Errors
+		/// * `AirdropDoesNotExist` - No Airdrop exist that is associated 'airdrop_id'
+		/// * `NotAirdropCreator` - Signer of the origin is not the creator of the Airdrop
+		fn disable_airdrop(
+			origin_id: Self::AccountId,
+			airdrop_id: Self::AirdropId,
+		) -> Result<Self::Balance, DispatchError> {
+			let airdrop = Self::get_airdrop(&airdrop_id)?;
+			ensure!(airdrop.creator == origin_id, Error::<T>::NotAirdropCreator);
+
+			let unclaimed_funds = Airdrops::<T>::try_mutate(airdrop_id, |airdrop| {
+				match airdrop.as_mut() {
+					Some(airdrop) => {
+						let at = T::Time::now();
+						let unclaimed_funds = airdrop.total_funds - airdrop.claimed_funds;
+
+						// REVIEW: Checking each recipient fund to see if they have started
+						// claiming could prove to be expensive. Should we instead require that all
+						// funds be claimed for an airdrop to end?
+						// sets claimed funds equal to total funds so the airdrop can be pruned
+						airdrop.disabled = true;
+						airdrop.claimed_funds = airdrop.total_funds;
+
+						Self::deposit_event(Event::AirdropEnded { airdrop_id, at });
+
+						Ok(unclaimed_funds)
+					},
+					None => Err(Error::<T>::AirdropDoesNotExist.into()),
+				}
+			});
+
+			Self::prune_airdrop(airdrop_id)?;
+
+			unclaimed_funds
+		}
+
+		/// Claim a recipient reward from an Airdrop.
+		///
+		/// # Errors
+		/// * `AirdropDoesNotExist` - No Airdrop exist that is associated 'airdrop_id'
+		/// * `AirdropIsNotEnabled` - The Airdrop has not been enabled
+		/// * `RecipientNotFound` - No recipient associated with the `identity` could be found.
+		fn claim(
+			airdrop_id: Self::AirdropId,
+			identity: Self::Identity,
+			reward_account: Self::AccountId,
+		) -> DispatchResultWithPostInfo {
+			let airdrop_account = Self::get_airdrop_account_id(airdrop_id);
+			let (available_to_claim, recipient_fund) =
+				RecipientFunds::<T>::try_mutate(airdrop_id, identity, |fund| {
+					match fund.as_mut() {
+						Some(fund) => {
+							let claimable = Self::claimable(airdrop_id, fund)?;
+							// .map_err(|_| Error::<T>::AirdropDoesNotExist)?;
+							let available_to_claim = claimable.saturating_sub(fund.claimed);
+
+							ensure!(
+								available_to_claim > T::Balance::zero(),
+								Error::<T>::NothingToClaim
+							);
+
+							// Update Airdrop and fund status
+							(*fund).claimed = fund.claimed.saturating_add(available_to_claim);
+
+							Ok((available_to_claim, *fund))
+						},
+						None => Err(Error::<T>::RecipientNotFound),
+					}
+				})?;
+
+			T::RecipientFundAsset::transfer(
+				&airdrop_account,
+				&reward_account,
+				available_to_claim,
+				false,
+			)?;
+
+			Airdrops::<T>::try_mutate(airdrop_id, |airdrop| match airdrop.as_mut() {
+				Some(airdrop) => {
+					airdrop.claimed_funds =
+						airdrop.claimed_funds.saturating_add(available_to_claim);
```

Comment:
is saturating arithmetic OK here? or should `checked_*`/`safe_*` be used

Reply:
f32f7b768a77d3c98121bac6081837f7bc685a84